File: ABC/276/E.cpp
#include <iostream>

using namespace std;
using ll = long long;

#define rep(i, n) for (int i = 0; i < (n); i++)

int h, w;
const int x[4] = {0, 0, 1, -1};
const int y[4] = {1, -1, 0, 0};
string g[1000001];
bool visited[1000001];

bool dfs(int r, int c, int d) {
  visited[r*w+c] = true;
  for (int i = 0; i < 4; i++) {
    int nr = r+x[i], nc = c+y[i];
    if (nr < 0 || nr >= h || nc < 0 || nc >= w || g[nr][nc] == '#') continue;
    if (g[nr][nc] == 'S' && d >= 4) return true;
    if (visited[nr*w+nc]) continue;
    if (dfs(nr, nc, d+1)) return true;
  }
  return false;
}

int main() {
  cin >> h >> w;
  rep(i, h) cin >> g[i];
  int sr = -1;
  int sc = -1;
  rep(i, h)rep(j, w) if (g[i][j] == 'S') {sr = i; sc = j;}
  cout << (dfs(sr, sc, 0) ? "Yes" : "No") << endl;
}

File: ABC/276/E.py
# TLE
from collections import deque
import sys
sys.setrecursionlimit(10**9)

h, w = map(int, input().split())
g = [list(input()) for _ in range(h)]
s = [[0] * w for _ in range(h)]
sr, sc = -1, -1

for i in range(h):
    for j in range(w):
        if g[i][j] == "S":
            sr, sc = i, j

x = [0, 0, 1, -1]
y = [1, -1, 0, 0]
visited = [[False] * w for _ in range(h)]


def dfs(r, c, d):
    if r < 0 or r >= h or c < 0 or c >= w or g[r][c] == "#":
        return False
    if r == sr and c == sc and d >= 4:
        return True
    if visited[r][c]:
        return False
    s[r][c] = d
    visited[r][c] = True
    res = False
    for dx, dy in zip(x, y):
        nr, nc = r+dx, c+dy
        if nr < 0 or nr >= h or nc < 0 or nc >= w or g[nr][nc] == "#":
            continue
        res |= dfs(nr, nc, d+1)
    return res


if dfs(sr, sc, 0):
    print("Yes")
else:
    print("No")
